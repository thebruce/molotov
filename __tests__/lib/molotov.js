const Molotov = require('../../lib/molotov');
const { MolotovError } = require('../../lib/molotovError');

const molotovConfig = {
  testMolotovImplementer: {
    supersNameSpace: {
      testSuper: 'testSuper'
    },
    molotovPlugins: {
      testSuper: {
        onlyPluginOne: [
          'pluginOne',
        ],
        pluginAll: [
          'pluginOne',
          'pluginTwo',
          'pluginThree',
        ],
        pluginOneTwo: [
          'pluginOne',
          'pluginTwo',
        ],
      },
    },
  },
};

const configOverrides = {
  testMolotovImplementer: {
    molotovPlugins: {
      testSuper: {
        pluginAll: [
          'pluginOne',
          'pluginTwo',
        ],
        pluginOneThree: [
          'pluginOne',
          'pluginThree',
        ],
      },
    },
  },
};

const mixins = require('./../helpers/mixins');
const supers = require('./../helpers/supers');
const cocktailClass = require('./../helpers/cocktailHelper');

let tmpMocks = [];

describe('Molotov Basic tests', () => {
  beforeEach(() => {
    tmpMocks.forEach(mock => mock.mockRestore());
    tmpMocks = [];
    jest.resetAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test('Molotov constructor', () => {
    expect.assertions(7);
    const molotov = new Molotov(molotovConfig, 'testMolotovImplementer', supers, mixins, configOverrides, [cocktailClass]);
    molotov.setPlugins({ plugins: 'yay' });
    expect(molotov.getNameSpace()).toEqual('testMolotovImplementer');
    expect(molotov.getMixins()).toMatchSnapshot();
    expect(molotov.getMolotovConfig()).toMatchSnapshot();
    expect(molotov.getPlugins()).toEqual({ plugins: 'yay' });
    expect(molotov.getSupers()).toMatchSnapshot();
    expect(molotov.getConfigOverrides()).toMatchSnapshot();
    expect(molotov.getCocktails()).toMatchSnapshot();
  });
});

describe('Molotov Constructor Errors', () => {
  beforeEach(() => {
    tmpMocks.forEach(mock => mock.mockRestore());
    tmpMocks = [];
    jest.resetAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test('Bad config', () => {
    expect.assertions(1);
    expect(() => new Molotov(null, 'testMolotovImplementer', supers, mixins, configOverrides, [cocktailClass])).toThrowError(MolotovError);
  });
  test('Config must not be string', () => {
    expect.assertions(1);
    expect(() => new Molotov('string', 'testMolotovImplementer', supers, mixins, configOverrides, [cocktailClass])).toThrowError(MolotovError);
  });
  test('Supers must be an object', () => {
    expect.assertions(1);
    expect(() => new Molotov(molotovConfig, 'testMolotovImplementer', 'supers', mixins, configOverrides, [cocktailClass])).toThrowError(MolotovError);
  });
  test('Mixins must have super keys', () => {
    expect.assertions(1);
    expect(() => new Molotov(molotovConfig, 'testMolotovImplementer', supers, {}, configOverrides, [cocktailClass])).toThrowError(MolotovError);
  });
  test('Mixins must be objects', () => {
    expect.assertions(1);
    expect(() => new Molotov(molotovConfig, 'testMolotovImplementer', supers, 'mixins', configOverrides, [cocktailClass])).toThrowError(MolotovError);
  });
  test('ConfigOverrides must be objects', () => {
    expect.assertions(1);
    expect(() => new Molotov(molotovConfig, 'testMolotovImplementer', supers, mixins, 'configOverrides', [cocktailClass])).toThrowError(MolotovError);
  });
  test('CocktailArray must be array', () => {
    expect.assertions(1);
    expect(() => new Molotov(molotovConfig, 'testMolotovImplementer', supers, mixins, configOverrides, 'cocktail')).toThrowError(MolotovError);
  });
  test('Bad Molotov config throws an error', () => {
    expect.assertions(1);
    expect(() => new Molotov({nob: 'nope'}, 'testMolotovImplementer', supers, mixins, configOverrides, 'cocktail')).toThrowError(MolotovError);
  });
});
