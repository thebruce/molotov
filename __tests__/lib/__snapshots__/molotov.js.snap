// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molotov Basic tests Molotov constructor 1`] = `
Object {
  "testSuper": Object {
    "pluginOne": [Function],
    "pluginThree": [Function],
    "pluginTwo": [Function],
  },
}
`;

exports[`Molotov Basic tests Molotov constructor 2`] = `
Object {
  "testMolotovImplementer": Object {
    "molotovPlugins": Object {
      "testSuper": Object {
        "onlyPluginOne": Array [
          "pluginOne",
        ],
        "pluginAll": Array [
          "pluginOne",
          "pluginTwo",
          "pluginThree",
        ],
        "pluginOneTwo": Array [
          "pluginOne",
          "pluginTwo",
        ],
      },
    },
    "supersNameSpace": Object {
      "testSuper": "testSuper",
    },
  },
}
`;

exports[`Molotov Basic tests Molotov constructor 3`] = `
Object {
  "testSuper": [Function],
}
`;

exports[`Molotov Basic tests Molotov constructor 4`] = `
Object {
  "testMolotovImplementer": Object {
    "molotovPlugins": Object {
      "testSuper": Object {
        "pluginAll": Array [
          "pluginOne",
          "pluginTwo",
        ],
        "pluginOneThree": Array [
          "pluginOne",
          "pluginThree",
        ],
      },
    },
  },
}
`;

exports[`Molotov Basic tests Molotov constructor 5`] = `
Array [
  CocktailHelper {
    "cocktailConfig": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "onlyPluginOne": Array [
              "pluginOne",
            ],
            "pluginAll": Array [
              "pluginFour",
            ],
            "pluginOneTwo": Array [
              "pluginOne",
              "pluginTwo",
              "pluginFour",
            ],
          },
          "testSuperTwo": Object {
            "pluginSeven": Array [
              "pluginSeven",
            ],
          },
        },
        "supersNameSpace": Object {
          "testSuper": "testSuper",
          "testSuperTwo": "testSuperTwo",
        },
      },
    },
    "cocktailMixins": Object {
      "testSuper": Object {
        "pluginFour": [Function],
        "pluginThree": [Function],
      },
      "testSuperTwo": Object {
        "pluginSeven": [Function],
      },
    },
    "cocktailSupers": Object {
      "testSuper": [Function],
      "testSuperTwo": [Function],
    },
  },
]
`;

exports[`Molotov advanced Molotov Mix Supers 1`] = `
SuperMixologist {
  "molotov": Molotov {
    "cocktails": Array [
      CocktailHelper {
        "cocktailConfig": Object {
          "testMolotovImplementer": Object {
            "molotovPlugins": Object {
              "testSuper": Object {
                "onlyPluginOne": Array [
                  "pluginOne",
                ],
                "pluginAll": Array [
                  "pluginFour",
                ],
                "pluginOneTwo": Array [
                  "pluginOne",
                  "pluginTwo",
                  "pluginFour",
                ],
              },
              "testSuperTwo": Object {
                "pluginSeven": Array [
                  "pluginSeven",
                ],
              },
            },
            "supersNameSpace": Object {
              "testSuper": "testSuper",
              "testSuperTwo": "testSuperTwo",
            },
          },
        },
        "cocktailMixins": Object {
          "testSuper": Object {
            "pluginFour": [Function],
            "pluginThree": [Function],
          },
          "testSuperTwo": Object {
            "pluginSeven": [Function],
          },
        },
        "cocktailSupers": Object {
          "testSuper": [Function],
          "testSuperTwo": [Function],
        },
      },
    ],
    "configOverrides": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "pluginAll": Array [
              "pluginOne",
              "pluginTwo",
            ],
            "pluginOneThree": Array [
              "pluginOne",
              "pluginThree",
            ],
          },
        },
      },
    },
    "mixins": Object {
      "testSuper": Object {
        "pluginOne": [Function],
        "pluginThree": [Function],
        "pluginTwo": [Function],
      },
    },
    "molotovConfig": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "onlyPluginOne": Array [
              "pluginOne",
            ],
            "pluginAll": Array [
              "pluginOne",
              "pluginTwo",
              "pluginThree",
            ],
            "pluginOneTwo": Array [
              "pluginOne",
              "pluginTwo",
            ],
          },
        },
        "supersNameSpace": Object {
          "testSuper": "testSuper",
          "testSuperTwo": "testSuperTwo",
        },
      },
    },
    "molotovNameSpace": "testMolotovImplementer",
    "supers": Object {
      "testSuper": [Function],
      "testSuperTwo": [Function],
    },
  },
  "target": "supersNameSpace",
  "type": "Supers",
}
`;

exports[`Molotov advanced Molotov Mix Supers 2`] = `
Polttopullo {
  "molotov": Molotov {
    "cocktails": Array [
      CocktailHelper {
        "cocktailConfig": Object {
          "testMolotovImplementer": Object {
            "molotovPlugins": Object {
              "testSuper": Object {
                "onlyPluginOne": Array [
                  "pluginOne",
                ],
                "pluginAll": Array [
                  "pluginFour",
                ],
                "pluginOneTwo": Array [
                  "pluginOne",
                  "pluginTwo",
                  "pluginFour",
                ],
              },
              "testSuperTwo": Object {
                "pluginSeven": Array [
                  "pluginSeven",
                ],
              },
            },
            "supersNameSpace": Object {
              "testSuper": "testSuper",
              "testSuperTwo": "testSuperTwo",
            },
          },
        },
        "cocktailMixins": Object {
          "testSuper": Object {
            "pluginFour": [Function],
            "pluginThree": [Function],
          },
          "testSuperTwo": Object {
            "pluginSeven": [Function],
          },
        },
        "cocktailSupers": Object {
          "testSuper": [Function],
          "testSuperTwo": [Function],
        },
      },
    ],
    "configOverrides": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "pluginAll": Array [
              "pluginOne",
              "pluginTwo",
            ],
            "pluginOneThree": Array [
              "pluginOne",
              "pluginThree",
            ],
          },
        },
      },
    },
    "mixins": Object {
      "testSuper": Object {
        "pluginOne": [Function],
        "pluginThree": [Function],
        "pluginTwo": [Function],
      },
    },
    "molotovConfig": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "onlyPluginOne": Array [
              "pluginOne",
            ],
            "pluginAll": Array [
              "pluginOne",
              "pluginTwo",
              "pluginThree",
            ],
            "pluginOneTwo": Array [
              "pluginOne",
              "pluginTwo",
            ],
          },
        },
        "supersNameSpace": Object {
          "testSuper": "testSuper",
          "testSuperTwo": "testSuperTwo",
        },
      },
    },
    "molotovNameSpace": "testMolotovImplementer",
    "supers": Object {
      "testSuper": [Function],
      "testSuperTwo": [Function],
    },
  },
  "target": "molotovPlugins",
  "type": "Plugins",
}
`;

exports[`Molotov no overrides no cocktails Molotov constructor 1`] = `
Object {
  "testSuper": Object {
    "pluginOne": [Function],
    "pluginThree": [Function],
    "pluginTwo": [Function],
  },
}
`;

exports[`Molotov no overrides no cocktails Molotov constructor 2`] = `
Object {
  "testSuper": [Function],
}
`;

exports[`Molotov no overrides no cocktails Molotov constructor 3`] = `
Polttopullo {
  "molotov": Molotov {
    "cocktails": Array [],
    "configOverrides": Object {},
    "mixins": Object {
      "testSuper": Object {
        "pluginOne": [Function],
        "pluginThree": [Function],
        "pluginTwo": [Function],
      },
    },
    "molotovConfig": Object {
      "testMolotovImplementer": Object {
        "molotovPlugins": Object {
          "testSuper": Object {
            "onlyPluginOne": Array [
              "pluginOne",
            ],
            "pluginAll": Array [
              "pluginOne",
              "pluginTwo",
              "pluginThree",
            ],
            "pluginOneTwo": Array [
              "pluginOne",
              "pluginTwo",
            ],
          },
        },
        "supersNameSpace": Object {
          "testSuper": "testSuper",
        },
      },
    },
    "molotovNameSpace": "testMolotovImplementer",
    "supers": Object {
      "testSuper": [Function],
    },
  },
  "target": "molotovPlugins",
  "type": "Plugins",
}
`;
